Enforce that the task calling `awaitRunningTasks()` doesn't belong to the pool

gkotian opened the pull request:
As promised in https://github.com/sociomantic-tsunami/ocean/pull/38#discussion_r106163458

nemanja-boric-sociomantic commented on relnotes/ipool.feature.md:
```diff
@@ -0,0 +1,7 @@
+* `ocean.util.container.pool.model.IPool`, `ocean.util.container.pool.model.IAggregatePool`
+
+  A new abstract function `isBusy()` has been added to the API of IPool. Classes
+  wishing to inherit from `IPool` must provide an implementation for it.
```
`type-breaking`?

gkotian replied:
Yep, it'll be a breaking change for any class that inherits from `IPool`. Does that mean that this PR should then target `v3.x.x`?

nemanja-boric-sociomantic replied:
Yes, it should target v3.x.x and the milestone would be v4.0.0. I wonder if this can be done differently (see my other comment), so we avoid breaking change.

mihails-strasuns-sociomantic replied:
Why don't you move this method to `IPool` itself, possibly even marking as final? Both `num_busy` and `getItemIndex` are available in it.

gkotian replied:
>Why don't you move this method to IPool itself, possibly even marking as final? Both num_busy and getItemIndex are available in it.

Ah yes indeed. I'm not sure why I didn't think of this simpler approach. Will try it out now.

nemanja-boric-sociomantic commented on src/ocean/task/TaskPool.d:
```diff
@@ -176,11 +176,14 @@ class TaskPool ( TaskT : Task ) : ObjectPool!(Task)
 
         foreach (task; tasks_iterator)
         {
+            if (!isBusy(this.toItem(task)))
```
`this.isBusy`?

nemanja-boric-sociomantic commented on src/ocean/task/TaskPool.d:
```diff
@@ -181,6 +185,9 @@ class TaskPool ( TaskT : Task ) : ObjectPool!(Task)
 
         foreach (task; tasks_iterator)
         {
+            enforce(!isSame(this.toItem(current_task), this.toItem(task)),
```
`this.`?

nemanja-boric-sociomantic replied:
I would go as far as asserting here?

gkotian replied:
>I would go as far as asserting here?

Does it make a difference here? Using `enforce` allows me to unit test it using `testThrown()`, but I guess I won't be able to do that if I use `assert` instead.

mihails-strasuns-sociomantic replied:
`assert` is not applicable here by our guidelines as there are no other safety checks against the situation and it will be removed in `-release`.

nemanja-boric-sociomantic replied:
Isn't it a programmer's error to end up in this situation? Is there any other way how would you end in this situation?

mihails-strasuns-sociomantic replied:
> Isn't it a programmer's error to end up in this situation?

Yes and it doesn't make your life any better when deployed app silently hangs in production. You may need to re-read https://github.com/sociomantic/backend/blob/master/doc/coding/assert-vs-enforce.rst :P

nemanja-boric-sociomantic replied:
Sure, I don't mind this, nor I will insist on `assert` here. I was just under impression that this check is very likely to fail the assertion during development time, where it can be safely removed from the production.

nemanja-boric-sociomantic commented:
I wonder if, instead of checking at the iteration time, you could perform a search to find out if the current task is contained in a pool?

nemanja-boric-sociomantic commented:
i.e. if it would be better to provide `IAggregatePool.isOwner` or similar method.

mihails-strasuns-sociomantic commented on src/ocean/util/container/pool/model/IAggregatePool.d:
```diff
@@ -639,14 +639,31 @@ public abstract class IAggregatePool ( T ) : IPool, IFreeList!(ItemType_!(T))
+
+    protected override bool isBusy ( Item item )
+    {
+        return this.getItemIndex(item) < this.num_busy_;
```
Is it even correct? Are items moved in the pool when they are recycled?

nemanja-boric-sociomantic replied:
looks like this might be correct, but it's implementation detail of `IPool` (one more vote for having it there: https://github.com/gkotian/ocean/blob/aab03da14b6012f85da7fc0cfe888a5878239f8e/src/ocean/util/container/pool/model/IPool.d#L413)

nemanja-boric-sociomantic replied:
This is where the shuffling of the last busy element and recycled element is done: https://github.com/gkotian/ocean/blob/aab03da14b6012f85da7fc0cfe888a5878239f8e/src/ocean/util/container/pool/model/IPool.d#L480

gkotian commented:
Updated with `isBusy()` moved to `IPool`.

```diff
diff --git a/relnotes/ipool.feature.md b/relnotes/ipool.feature.md
index e1ea926..5470124 100644
--- a/relnotes/ipool.feature.md
+++ b/relnotes/ipool.feature.md
@@ -1,7 +1,4 @@
-* `ocean.util.container.pool.model.IPool`, `ocean.util.container.pool.model.IAggregatePool`
+* `ocean.util.container.pool.model.IPool`
 
-  A new abstract function `isBusy()` has been added to the API of IPool. Classes
-  wishing to inherit from `IPool` must provide an implementation for it.
-
-  The `IAggregatePool` class (which itself inherits from `IPool`) has been
-  updated to provide its implementation of the `isBusy()` method.
+  A new function `isBusy()` has been added to the API of IPool. It can be used
+  to check whether a given pool item is currently busy or idle.
diff --git a/src/ocean/util/container/pool/model/IAggregatePool.d b/src/ocean/util/container/pool/model/IAggregatePool.d
index 22bf0c1..f5605d3 100644
--- a/src/ocean/util/container/pool/model/IAggregatePool.d
+++ b/src/ocean/util/container/pool/model/IAggregatePool.d
@@ -648,29 +648,12 @@ public abstract class IAggregatePool ( T ) : IPool, IFreeList!(ItemType_!(T))
 
     /***************************************************************************
 
-        Checks if item is currently busy.
-
-        Params:
-            item = item to check
-
-        Returs:
-            true if item is currently busy, false otherwise.
-
-    ***************************************************************************/
-
-    protected override bool isBusy ( Item item )
-    {
-        return this.getItemIndex(item) < this.num_busy_;
-    }
-
-    /***************************************************************************
-
         Checks if item is null.
 
diff --git a/src/ocean/util/container/pool/model/IPool.d b/src/ocean/util/container/pool/model/IPool.d
index 696bced..6031190 100644
--- a/src/ocean/util/container/pool/model/IPool.d
+++ b/src/ocean/util/container/pool/model/IPool.d
@@ -324,6 +324,23 @@ public abstract class IPool : IPoolInfo, ILimitable
 
     /***************************************************************************
 
+        Checks if item is currently busy.
+
+        Params:
+            item = item to check
+
+        Returns:
+            true if item is currently busy, false otherwise.
+
+    ***************************************************************************/
+
+    public bool isBusy ( Item item )
+    {
+        return this.getItemIndex(item) < this.num_busy_;
+    }
+
+    /***************************************************************************
+
         Ensures that the pool contains at least the specified number of items.
         Useful to pre-allocate a pool of a certain size.
 
@@ -558,26 +575,12 @@ public abstract class IPool : IPoolInfo, ILimitable
 
     /***************************************************************************
 
-        Checks if item is currently busy.
-
-        Params:
-            item = item to check
-
-        Returs:
-            true if item is currently busy, false otherwise.
-
-    ***************************************************************************/
-
-    abstract protected bool isBusy ( Item item );
-
-    /***************************************************************************
-
         Checks if item is null.
 
```

gkotian commented:
With this, I suppose `type-breaking` can be removed.

Also, @nemanja-boric-sociomantic, is your comment about having a `IAggregatePool.isOwner` or similar method still applicable?

nemanja-boric-sociomantic commented:
> Also, @nemanja-boric-sociomantic, is your comment about having a IAggregatePool.isOwner or similar method still applicable?

No, now it's better, that was just one suggestion to avoid breaking change.

nemanja-boric-sociomantic commented on src/ocean/task/TaskPool.d:
```diff
@@ -181,6 +185,9 @@ class TaskPool ( TaskT : Task ) : ObjectPool!(Task)
 
         foreach (task; tasks_iterator)
         {
+            enforce(!this.isSame(this.toItem(current_task), this.toItem(task)),
```
So, this works only with the `AllItemsIterator`, but not with `BusyItemsIterator`, right?

mihails-strasuns-sociomantic replied:
Yeah, this PR introduces minor performance pessimization to verify the sanity condition. Should be fine for its use case but worth keeping in mind (maybe even put an explicit comment about it, @gkotian ?)

gkotian replied:
Hah, no. In theory, I could've simply used `BusyItemsIterator` (and not even had to define the new `isBusy` method), but I just wanted to play it safe and deliberately iterate over all tasks in the pool, including the idle ones.

nemanja-boric-sociomantic replied:
That's ok, but I don't understand why iteration over all tasks is necessary in this context. Because if the current task is part of the task pool, it must be non-idle element, or?

nemanja-boric-sociomantic replied:
> Hah, no. In theory, I could've simply used BusyItemsIterator (and not even had to define the new isBusy method), but I just wanted to play it safe and deliberately iterate over all tasks in the pool, including the idle ones.

Yeah, I was trying to figure this out :-). I will defer to @mihails-strasuns-sociomantic here as I don't have Task profiling experience to be able to figure out how big the ratio of busy/all items is.

mihails-strasuns-sociomantic replied:
Pedantic approach would be to iterate over `BusyItems` first and `enforce`, and after that call a delegate from within `assert` that iterates over `IdleItems` only. Latter indeed is not possible if rest of the scheduler is in sane state but can serve as an additional debugging helper (== ``assert`).

mihails-strasuns-sociomantic replied:
(I don't mind either of "current state" vs "only busy items" vs "enforce busy, assert idle" options)

nemanja-boric-sociomantic replied:
Me neither, I was just trying to understand the previous commit's message :-).

mihails-strasuns-sociomantic commented on src/ocean/task/TaskPool.d:
```diff
@@ -361,3 +371,44 @@ unittest
+
+        override public void run ( )
+        {
+            this.non_throwing_dg();
```
Why this no-op delegate call is needed at all? Looks confusing.

mihails-strasuns-sociomantic replied:
Also can copy pool reference instead of delegate and call `pool.awaitRunningTasks` directly from `run`

gkotian replied:
>Why this no-op delegate call is needed at all? Looks confusing.

OK, I'll get rid of it.

>Also can copy pool reference instead of delegate and call `pool.awaitRunningTasks` directly from `run`

How can I copy it without running into the forward reference errors?
That is, if I do this:

```d
unittest
{
    // Test that 'awaitRunningTasks()' cannot be called from a task that itself
    // belongs to the pool.

    static class AwaitTask : Task
    { 
        private TaskPool!(AwaitTask) pool;

        public void copyArguments ( TaskPool!(AwaitTask) pool ) 
        { 
            this.pool = pool;
        } 

        override public void run ( ) 
        { 
            this.pool.awaitRunningTasks();
        } 
    } 

    auto pool = new TaskPool!(AwaitTask);
    initScheduler(SchedulerConfiguration.init);

    pool.start(pool);
}
```

it fails with:
```
Error: template instance ocean.core.Traits_tango.ParameterTupleOf!(copyArguments) forward reference of copyArguments
```

mihails-strasuns-sociomantic replied:
I see, didn't expect that to be an issue. Nevermind in that case, LGTM

mihails-strasuns-sociomantic reviewed (commented):
Few minor comments on test case

gkotian commented:
Removed the extra non-throwing delegate.

```diff
@@ -379,21 +379,16 @@ unittest
 
     static class AwaitTask : Task
     {
-        void delegate () non_throwing_dg;
-        void delegate () throwing_dg;
+        void delegate () dg;
 
-        public void copyArguments ( void delegate () non_throwing_dg,
-            void delegate () throwing_dg )
+        public void copyArguments ( void delegate () dg )
         {
-            this.non_throwing_dg = non_throwing_dg;
-            this.throwing_dg = throwing_dg;
+            this.dg = dg;
         }
 
         override public void run ( )
         {
-            this.non_throwing_dg();
-
-            testThrown!(Exception)(this.throwing_dg());
+            testThrown!(Exception)(this.dg());
         }
     }
 
@@ -402,12 +397,9 @@ unittest
 
     pool.start(
         {
-            // This is a non throwing delegate as it does nothing.
-        },
-        {
-            // On the other hand, this delegate should throw as it will attempt
-            // to call the task pool's awaitRunningTasks() method from within a
-            // task that itself belongs to the pool.
+            // This delegate should throw as it will attempt to call the task
+            // pool's awaitRunningTasks() method from within a task that itself
+            // belongs to the pool.
             pool.awaitRunningTasks();
         }
     );
```

gkotian committed:
Minor formatting & typo fixes

gkotian committed:
Add an 'isBusy()' function to IPool

This function will return true if the given pool item is currently busy,
and false if it is currently idle.

gkotian committed:
Enforce current task is not from the task pool

The 'awaitRunningTasks()' function can wait for all running tasks from
the pool to finish only if the current task that calls it doesn't itself
belong to the pool. Enforce that this is always the case.

gkotian committed:
Iterate over all pool items in awaitRunningTasks

In the 'awaitRunningTasks()' function, instead of iterating over only
the busy tasks in the task pool, iterate over all tasks instead. But
continue the iteration loop immediately if the task being iterated upon
is discovered to be busy.

This change retains the exact same semantics, but will allow us to
perform other checks in future.

mihails-strasuns-sociomantic commented:
Thanks!
